README.md
[![Build Status](https://jenkins.grotax.de/buildStatus/icon?job=FFSH-Firmware)](https://jenkins.grotax.de/job/FFSH-Firmware/)

# Freifunk Südholstein Site-Konfiguration

In diesem Repository wird die Gluon Konfiguration für Freifunk Südholstein gepflegt.


## Branch Struktur

| branch  | gluon-branch  | Kommentar                                          |
|---------|---------------|----------------------------------------------------|
| dev     | master        | Nur für Entwickler enthält experimentelle Software |
| testing | v2018.1.x     | Zum testen einer neuen Version für interessierte.  |
| rc      | v2018.1 (tag) | Zum erneuten testen (in einer größeren Gruppe)     |
| stable  | v2018.1 (tag) | Für alle anderen, empfohlene Version               |


## Jenkins
Die Firmware wird automatisch mit der Hilfe von build.py gebaut. Sobald ein commit in das Repository gepusht wird informiert GitHub die Jenkins Instaz über die Änderung und Jenkins prüft ob ein build notwendig ist.
Wenn du änderungen am Repository machen willst die keinen build auslösen sollen (dieser benötigt 1-3 Stunden) dann kannst du deiner Commit Message ein [NO-BUILD] als Prefix verpassen.
```
git commit -m "[NO-BUILD] änderungen am changelog"

```
Den aktuellen build Status siehst du oben in dieser README oder du gehst auf [jenkins.grotax.de](https://jenkins.grotax.de). Wenn der Build erfolgreich war, werden die images auf [firmware.grotax.de](https://firmware.grotax.de) veröffentlicht. Diese sind jedoch nicht für das automatische Update freigegeben.

## Firmware selber bauen:

Für das bauen der Firmware haben wir ein Python Script. Dadurch können wir die Firmware automatisch bauen.
Das script kann aber auch zum manuellen bauen genutzt werden. Ein Beispiel findet sich weiter unten.

### build.py Argumente:

Hier werden alle verfügbaren Befehle des Scripts erläutert.

| command | value    | make "equivalent" | Kommentar                                       |
|---------|----------|-------------------|-------------------------------------------------|
| -c      | update   | make update       | lädt opwenwrt und wendet gluon patches an       |
| -c      | build    | make build        | baut die firmware                               |
| -c      | clean    | make clean        | löscht alle packages des targets                |
| -c      | dirclean | make dirclean     | löscht alle targets und die toolchain (kaputt!) |
| -c      | sign     | n/a               | signiert die firmware                           |

Weitere Argumente für das Script:

| command | value | default | name | pflicht | Kommentar |
|---|---|---|---|---|----|
| -b | dev or testing or rc or stable | dev | Branch | ja | der Firmware branch |
| -w | site | n/a | Workspace | ja | Pfad zum site Repository |
| -n | 42 | n/a | Build Number | ja | build Nummer wird von jenkins automatisch hochgezählt wird im firmware Namen verwendet |
| -t | ar71xx-generic or ... | all targets | Target | nein | ohne Angabe werden alle Targets gebaut, mit angabe nur der angegebene Target |
| -s | <pfad zu secret> | n/a | Secret | nein | wird nur beim signieren benötigt |
| -d | <pfad zu public directory> | n/a | Directory | nein | wird nur bei -publish benötigt |
| --commit | der verwendete commit | n/a | Commit | ja | commit sha, dient als Refferenz im build.json |
| --cores | 1 bis N | 1 | Cores | nein | Anzahl der zu verwenden Threads, Empfehlung: CPU-Kerne+1 |
| --log | w or s | s | Log | nein | Log level w: nur warnungen/Fehler, s: alles |

Beispiel:
```
./build.py -c update -b dev -n 42 -w $(pwd) --commit $(git rev-parse HEAD)
./build.py -c build -t "ar71xx-tiny" -b dev -n 42 -w $(pwd) --commit $(git rev-parse HEAD)
```


CHANGELOG.md
## Changelog:

---

#### 2018.1 VORLÄUFIG

| Branch  | Verfügbarkeit | Datum    |
|---------|---------------|----------|
| dev     | ja            | 11.07.18 |
| testing | ja            | 11.07.18 |
| rc      | nein          | 24.07.18 |
| stable  | nein          | 24.07.18 |

##### FFSH:

+ Zusammenführung von FFOD und FFRZ zu FFSH
+ Domains: ffbo, ffrz, ffsh
  + | Kürzel | Name                  | ssid                    |
    |--------|-----------------------|-------------------------|
    | ffrz   | Freifunk Lauenburg    | lauenburg.freifunk.net  |
    | ffod   | Freifunk Stormarn     | stormarn.freiunk.net    |
    | ffsh   | Freifunk Südholstein  | freifunk-südholstein.de |

Gemeinsame Firmware für Freifunk Stormarn und Lauenburg unter dem neuen Namen Freifunk Südholstein.
Die SSIDs bleiben über die Domain-Funktion erhalten, bei der Einrichtung kann eine Domain ausgewählt werden.

Um die Domains der Knoten nicht zu ändern wird es zunächst für jede community einen anderen release geben.


##### Geräte:
Mit dem neuen Release wurden einige neue Geräte hinzugefügt, die tatsächlich von uns unterstützten Geräte findest du auf.
https://freifunk-suedholstein.de/firmware

+ ALFA NETWORK
    + AP121F
+ AVM
    + FRITZ!Box 4020
+ OpenMesh
    + A40
    + A60
    + OM2P v4
    + OM2P-HS v4
+ TP-Link
    + Archer C59 v1
    + CPE210 v2
    + TL-WA901ND v5
    + Archer C2600
+ ZyXEL
    + NBG6716
+ GL Innovations
    + GL-MT300A
    + GL-MT300N
    +GL-MT750
+ VoCore
    + VoCore 2
    + VoCore (8MB, 16MB)
+ D-Link
    + DIR615 (D1, D2, D3, D4, H1)
+ A5
    + V11
+ LeMaker
    + Lamobo r1

---

build.sh
#!/bin/bash -e
# ======================================================================
# Build script for Freifunk Suedholstein firmware runningn on Jenkins CI
#
# Source: https://github.com/ffsh
# Web: https://freifunk-suedholstein.de
#
# Credits:
#   - Freifunk Fulda:        Freifunk Darmstadt for your great support.
#   - Freifunk Suedholstein: Freifunk Fulda for this nice script!
# ======================================================================

# Default make options
MAKEOPTS="V=s -j$(expr $(nproc) + 1)"

# Default to build all Gluon targets if parameter -t is not set
TARGETS="ar71xx-generic ar71xx-tiny ar71xx-nand brcm2708-bcm2708 brcm2708-bcm2709 mpc85xx-generic ramips-mt7621 sunxi-cortexa7 x86-generic x86-geode x86-64 ramips-mt7620 ramips-mt76x8 ramips-rt305x"

# Default is set to use current work directory
SITEDIR="$(pwd)"

# Default build identifier set to snapshot
BUILD="testing"

# Specify deployment server and user
DEPLOYMENT_SERVER="firmware.freifunk-suedholstein.de"
DEPLOYMENT_USER="deployment"

# Path to signing key
SIGNKEY=""

# Error codes
E_ILLEGAL_ARGS=126

# Help function used in error messages and -h option
usage() {
  echo ""
  echo "Build script for Freifunk-Suedholstein gluon firmware."
  echo ""
  echo "-b: Firmware branch name (e.g. development)"
  echo "    Default: current git branch"
  echo "-c: Build command: update | clean | download | build | sign | upload | prepare"
  echo "-d: Enable bash debug output"
  echo "-h: Show this help"
  echo "-m: Setting for make options (optional)"
  echo "    Default: \"${MAKEOPTS}\""
  echo "-n: Build identifier (optional)"
  echo "    Default: \"${BUILD}\""
  echo "-t: Gluon targets architectures to build"
  echo "    Default: \"${TARGETS}\""
  echo "-r: Release number (optional)"
  echo "    Default: fetched from release file"
  echo "-w: Path to site directory"
  echo "    Default: current working directory"
  echo "-s: Path to signing key"
  echo "    Default: empty"
}

# Evaluate arguments for build script.
if [[ "${#}" == 0 ]]; then
  usage
  exit ${E_ILLEGAL_ARGS}
fi

# Evaluate arguments for build script.
while getopts b:c:dhm:n:t:w:s: flag; do
  case ${flag} in
    b)
        BRANCH="${OPTARG}"
        ;;
    c)
      case "${OPTARG}" in
        update)
          COMMAND="${OPTARG}"
          ;;
        clean)
          COMMAND="${OPTARG}"
          ;;
        download)
          COMMAND="${OPTARG}"
          ;;
        build)
          COMMAND="${OPTARG}"
          ;;
        sign)
          COMMAND="${OPTARG}"
          ;;
        upload)
          COMMAND="${OPTARG}"
          ;;
        prepare)
          COMMAND="${OPTARG}"
          ;;
        *)
          echo "Error: Invalid build command set."
          usage
          exit ${E_ILLEGAL_ARGS}
          ;;
      esac
      ;;
    d)
      set -x
      ;;
    h)
      usage
      exit
      ;;
    m)
      MAKEOPTS="${OPTARG}"
      ;;
    n)
      BUILD="${OPTARG}"
      ;;
    t)
      TARGETS="${OPTARG}"
      ;;
    r)
      RELEASE="${OPTARG}"
      ;;
    w)
      # Use the root project as site-config for make commands below
      SITEDIR="${OPTARG}"
      ;;
    s)
      SIGNKEY="${OPTARG}"
      ;;
    *)
      usage
      exit ${E_ILLEGAL_ARGS}
      ;;
  esac
done

# Strip of all remaining arguments
shift $((OPTIND - 1));

# Check if there are remaining arguments
if [[ "${#}" > 0 ]]; then
  echo "Error: To many arguments: ${*}"
  usage
  exit ${E_ILLEGAL_ARGS}
fi

# Set branch name
if [[ -z "${BRANCH}" ]]; then
  BRANCH=$(git symbolic-ref -q HEAD)
  BRANCH=${BRANCH##refs/heads/}
  BRANCH=${BRANCH:-HEAD}
fi

# Set command
if [[ -z "${COMMAND}" ]]; then
  echo "Error: Build command missing."
  usage
  exit ${E_ILLEGAL_ARGS}
fi

# Set release number
if [[ -z "${RELEASE}" ]]; then
  RELEASE=$(cat "${SITEDIR}/release")
fi

# Normalize the branch name
BRANCH="${BRANCH#origin/}" # Use the current git branch as autoupdate branch
BRANCH="${BRANCH//\//-}"   # Replace all slashes with dashes

# Get the GIT commit description
COMMIT="$(git describe --always --dirty)"

# Number of days that may pass between releasing an updating
PRIORITY=1

update() {
  make ${MAKEOPTS} \
       GLUON_SITEDIR="${SITEDIR}" \
       GLUON_OUTPUTDIR="${SITEDIR}/output" \
       GLUON_RELEASE="${RELEASE}-${BUILD}" \
       GLUON_BRANCH="${BRANCH}" \
       GLUON_PRIORITY="${PRIORITY}" \
       update
}

clean() {
  for TARGET in ${TARGETS}; do
    echo "--- Update Gluon Dependencies for target: ${TARGET}"
    make ${MAKEOPTS} \
         GLUON_SITEDIR="${SITEDIR}" \
         GLUON_OUTPUTDIR="${SITEDIR}/output" \
         GLUON_RELEASE="${RELEASE}-${BUILD}" \
         GLUON_BRANCH="${BRANCH}" \
         GLUON_PRIORITY="${PRIORITY}" \
         GLUON_TARGET="${TARGET}" \
         clean
  done
}

download() {
  for TARGET in ${TARGETS}; do
    echo "--- Download Gluon Dependencies for target: ${TARGET}"
    make ${MAKEOPTS} \
         GLUON_SITEDIR="${SITEDIR}" \
         GLUON_OUTPUTDIR="${SITEDIR}/output" \
         GLUON_RELEASE="${RELEASE}-${BUILD}" \
         GLUON_BRANCH="${BRANCH}" \
         GLUON_PRIORITY="${PRIORITY}" \
         GLUON_TARGET="${TARGET}" \
         download
  done
}

build() {
  for TARGET in ${TARGETS}; do
    echo "--- Build Gluon Images for target: ${TARGET}"
    case "${BRANCH}" in
      stable| \
      testing| \
      development)
        make ${MAKEOPTS} \
             GLUON_SITEDIR="${SITEDIR}" \
             GLUON_OUTPUTDIR="${SITEDIR}/output" \
             GLUON_RELEASE="${RELEASE}-${BUILD}" \
             GLUON_BRANCH="${BRANCH}" \
             GLUON_PRIORITY="${PRIORITY}" \
             GLUON_TARGET="${TARGET}" \
             all
        ;;

      *)
        make ${MAKEOPTS} \
             GLUON_SITEDIR="${SITEDIR}" \
             GLUON_OUTPUTDIR="${SITEDIR}/output" \
             GLUON_RELEASE="${RELEASE}-${BUILD}" \
             GLUON_BRANCH="${BRANCH}" \
             GLUON_TARGET="${TARGET}" \
             all
      ;;
    esac
  done

  echo "--- Build Gluon Manifest"
  make ${MAKEOPTS} \
       GLUON_SITEDIR="${SITEDIR}" \
       GLUON_OUTPUTDIR="${SITEDIR}/output" \
       GLUON_RELEASE="${RELEASE}-${BUILD}" \
       GLUON_BRANCH="${BRANCH}" \
       GLUON_PRIORITY="${PRIORITY}" \
       manifest

  echo "--- Write Build file"
  cat > "${SITEDIR}/output/images/build" <<EOF
DATE=$(date '+%Y-%m-%d %H:%M:%S')
VERSION=$(cat "${SITEDIR}/release")
RELEASE=${RELEASE}
BUILD=${BUILD}
BRANCH=${BRANCH}
COMMIT=${COMMIT}
HOST=$(uname -n)
EOF
}

sign() {
  echo "--- Sign Gluon Firmware Build"

  # Add the signature to the local manifest
  contrib/sign.sh \
      "${SIGNKEY}" \
      "${SITEDIR}/output/images/sysupgrade/${BRANCH}.manifest"
}

upload() {
  echo "--- Upload Gluon Firmware Images and Manifest"

  # Build the ssh command to use
  SSH="ssh"
  SSH="${SSH} -o stricthostkeychecking=no"
  SSH="${SSH} -p 22022"

  # Determine upload target prefix
  case "${BRANCH}" in
    stable| \
    testing| \
    development)
      TARGET="${BRANCH}"
      ;;

    *)
      TARGET="others/${BRANCH}"
      ;;
  esac

  # Create the target directory on server
  ${SSH} \
      ${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER} \
      -- \
      mkdir \
          --parents \
          --verbose \
          "firmware/${TARGET}/${RELEASE}-${BUILD}"

  # Copy images to server
  rsync \
      --verbose \
      --recursive \
      --compress \
      --progress \
      --links \
      --rsh="${SSH}" \
      "${SITEDIR}/output/images/" \
      "${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER}:firmware/${TARGET}/${RELEASE}-${BUILD}"

  # Link latest upload in target to 'current'
  ${SSH} \
      ${DEPLOYMENT_USER}@${DEPLOYMENT_SERVER} \
      -- \
      ln \
          --symbolic \
          --force \
          --no-target-directory \
          "${RELEASE}-${BUILD}" \
          "firmware/${TARGET}/current"
}

prepare() {
  echo "--- Prepare directory for upload"

  # Determine upload target prefix
  case "${BRANCH}" in
    stable| \
    testing| \
    development)
      TARGET="${BRANCH}"
      ;;

    *)
      TARGET="others/${BRANCH}"
      ;;
  esac

  # Create the target directory on server
  mkdir \
    --parents \
    --verbose \
    "${SITEDIR}/output/firmware/${TARGET}"

  # Copy images to directory
  mv \
    --verbose \
    "${SITEDIR}/output/images" \
    "${SITEDIR}/output/firmware/${TARGET}/${RELEASE}-${BUILD}"

  # Link latest upload in target to 'current'
  cd "${SITEDIR}/output"
  ln \
      --symbolic \
      --force \
      --no-target-directory \
      "${RELEASE}-${BUILD}" \
      "firmware/${TARGET}/current"
}

(
  # Change working directory to gluon tree
  cd "${SITEDIR}/gluon"

  # Execute the selected command
  ${COMMAND}
)


build.py
#!/usr/bin/python3
"""
Build script for Freifunk Suedholstein firmware runningn on Jenkins CI

Source: https://github.com/ffsh/site
Web: https://freifunk-suedholstein.de

"""
import argparse
import json
import os
import time
from datetime import datetime
import subprocess as sp

DEFAULTS = {
    'targets' : ['ar71xx-generic',
                 'ar71xx-tiny',
                 'ar71xx-nand',
                 'brcm2708-bcm2708',
                 'brcm2708-bcm2709',
                 'mpc85xx-generic',
                 'ramips-mt7621',
                 'sunxi',
                 'x86-generic',
                 'x86-geode',
                 'x86-64',
                 'ipq806x',
                 'ramips-mt7620',
                 'ramips-mt7628',
                 'ramips-rt305x'
                ],
    'gluon_dir': '/gluon',
    'make_cores' : "1",
    'make_loglevel': 'V=s',
    'release': '2018.1',
    'priority': 1,
    'branch': 'dev'
}

PARSER = argparse.ArgumentParser()
PARSER.add_argument("-c", metavar="Command", dest="command",
                    help="build.py -c clean | dirclean | update | build | sign | publish",
                    required=True)
PARSER.add_argument("-b", metavar="Branch", dest="branch",
                    help="build.py -b dev | testing | rc | stable", required=True)
PARSER.add_argument("-w", metavar="Workspace", dest="workspace",
                    help="build.py -w \"${WORKSPACE}\"", required=True)
PARSER.add_argument("-n", metavar="Build Number", dest="build_number",
                    help="build.py -n ${BUILD_NUMBER}", required=True)
PARSER.add_argument("-t", metavar="Target", dest="target",
                    help="build.py -t ar71xx-generic | ...", required=False)
PARSER.add_argument("-s", metavar="Secret", dest="secret",
                    help="build.py -s <path to secret>", required=False)
PARSER.add_argument("-d", metavar="Public Direcotry", dest="directory",
                    help="build.py -d /var/www/firmware (jenkins needs rw)", required=False)
PARSER.add_argument("--commit", metavar="Commit", dest="commit",
                    help="build.py --commit sha", required=True)
PARSER.add_argument("--cores", metavar="Cores", dest="cores",
                    help="build.py --cores 4", required=False)
PARSER.add_argument("--log", metavar="Log Level", dest="log",
                    help="build.py --log V=s (stdout+stderr) | V=w (warnings/errors)",
                    required=False)


ARGS = PARSER.parse_args()


def clean():
    """
    Cleans the output Directory. Not necessary!
    """
    try:
        target = str(ARGS.target)
        print(target+" "+type(target))
        all_targets = False
    except TypeError:
        all_targets = True

    if all_targets:
        for target in DEFAULTS["targets"]:
            print("Cleaning target: {}".format(target))
            sp.check_call(["make", "-C", ARGS.workspace+DEFAULTS['gluon_dir'],
                           "GLUON_SITEDIR="+ARGS.workspace,
                           "GLUON_TARGET="+target,
                           "clean"])
    else:
        sp.check_call(["make", "-C", ARGS.workspace+DEFAULTS['gluon_dir'],
                       "GLUON_SITEDIR="+ARGS.workspace,
                       "GLUON_TARGET="+target,
                       "clean"])

def dirclean():
    """
    Clean with dirclean
    """
    print("Starting dirclean ...")
    sp.check_call(["make", "-C", ARGS.workspace+DEFAULTS['gluon_dir'],
                   "GLUON_SITEDIR="+ARGS.workspace,
                   "dirclean"])
    print("dirclean done.")

def update():
    """
    Updates the repository
    """
    print("Starting update ...")
    sp.check_call(["make", "-C", ARGS.workspace+DEFAULTS['gluon_dir'],
                   "GLUON_SITEDIR="+ARGS.workspace,
                   "update"])
    print("Update done.")


def build():
    """
    Build images definded by -t or default targets
    """
    print("Info: Starting building ...")
    print("Info: delete OLD images in workdir...")
    dir_source = "{}/output/images/".format(ARGS.workspace)
    sp.check_call(["rm", "-rf", dir_source])
    build_errors = {
        "number": 0,
        "errors": []
    }
    if ARGS.target is not None:
        all_targets = False
        target = str(ARGS.target)
        print("Info: single target found")
    else:
        all_targets = True
        print("Info: build all Targets")

    if all_targets:
        for target in DEFAULTS["targets"]:
            print("Info: Building target: {}".format(target))
            try:
                sp.check_call(["make", "-j", DEFAULTS['make_cores'], "-C",
                               ARGS.workspace+DEFAULTS['gluon_dir'],
                               DEFAULTS['make_loglevel'],
                               "GLUON_SITEDIR="+ARGS.workspace,
                               "GLUON_RELEASE={}-{}-{}".format(DEFAULTS['release'],
                                                               ARGS.build_number,
                                                               DEFAULTS['branch']),
                               "GLUON_BRANCH="+DEFAULTS['branch'],
                               "GLUON_OUTPUTDIR={}/output".format(ARGS.workspace),
                               "GLUON_TARGET="+target,
                               "all"])
            except sp.CalledProcessError as process_error:
                print(process_error)
                build_errors["errors"].append(process_error)
                build_errors["number"] += 1

    else:
        print("Info: Building target: {}".format(target))
        sp.check_call(["make", "-j", DEFAULTS['make_cores'], "-C",
                       ARGS.workspace+DEFAULTS['gluon_dir'],
                       DEFAULTS['make_loglevel'],
                       "GLUON_SITEDIR="+ARGS.workspace,
                       "GLUON_RELEASE={}-{}-{}".format(DEFAULTS['release'],
                                                       ARGS.build_number,
                                                       DEFAULTS['branch']),
                       "GLUON_BRANCH="+DEFAULTS['branch'],
                       "GLUON_OUTPUTDIR={}/output".format(ARGS.workspace),
                       "GLUON_TARGET="+target,
                       "all"])

    print("Info: Generating buid.json")
    time_stamp_sec = time.time()
    time_stamp = datetime.fromtimestamp(time_stamp_sec).strftime('%Y-%m-%d-%H-%M-%S')
    data = {
        'build_date' : time_stamp,
        'release' : "{}-{}-{}".format(DEFAULTS['release'], DEFAULTS['branch'],
                                      ARGS.build_number),
        'branch' : DEFAULTS['branch'],
        'commit' : ARGS.commit
    }
    with open("{}/output/images/build.json".format(ARGS.workspace), "w") as file:
        json.dump(data, file)

    # Create manifest
    print("Createing Manifest ...")

    if not os.path.isdir("{}/tmp/origin/".format(ARGS.workspace)):
        # Make sure that the tmp dir exists
        sp.check_call(["mkdir", "-p", "{}/tmp/origin/".format(ARGS.workspace)])

    sp.check_call(["make", "-C", ARGS.workspace+DEFAULTS['gluon_dir'],
                   "GLUON_SITEDIR="+ARGS.workspace,
                   "GLUON_RELEASE={}-{}-{}".format(DEFAULTS['release'],
                                                   ARGS.build_number,
                                                   DEFAULTS['branch']),
                   "GLUON_BRANCH="+DEFAULTS['branch'],
                   "GLUON_OUTPUTDIR={}/output".format(ARGS.workspace),
                   "GLUON_TARGET="+target,
                   "manifest"])
    if build_errors["number"] > 0:
        print("=============================")
        print("Error: We have {} Build errors".format(build_errors["number"]))
        for error in build_errors["errors"]:
            print("-------------------------")
            print(error)
            print("-------------------------")
        print("=============================")
        exit(1)

def sign():
    """
    Signs the manifest
    """
    sp.check_call(["{}/gluon/contrib/sign.sh".format(ARGS.workspace), ARGS.secret,
                   "{}/output/images/sysupgrade/{}.manifest".format(ARGS.workspace,
                                                                    DEFAULTS['branch'])])

def publish():
    """
    Copy the images to a public directory
    """
    try:
        directory = str(ARGS.directory)
    except ValueError:
        raise ValueError("You need to provide a valid path eg. -d /var/www/firmware")
    if os.path.isdir(directory):
        # direcotry/BRANCH/build.json
        if os.path.isdir("{}/{}".format(directory, DEFAULTS['branch'])):

            print("Detected old builds move to archive")
            dir_source = "{}/{}".format(directory, DEFAULTS['branch'])

            with open(dir_source+"/build.json") as file:
                old_build_date = json.load(file)["build_date"]
            dir_target = "{}/archive/{}-{}".format(directory, DEFAULTS['branch'], old_build_date)
            sp.check_call(["mkdir", "-p", dir_target])
            sp.check_call(["rsync", "-Ltr", "--remove-source-files", dir_source, dir_target])

        dir_source = "{}/output/images/".format(ARGS.workspace)
        dir_target = "{}/{}".format(directory, DEFAULTS['branch'])
        sp.check_call(["rsync", "-Ltr", dir_source, dir_target])

        print("delete images in workdir...")
        sp.check_call(["rm", "-rf", dir_source])
    else:
        raise ValueError("{} Path does not exist!".format(directory))
    #clean()

def main():
    """
    Entry function
    """
    global DEFAULTS
    with open(ARGS.workspace + '/release.json', 'r') as file:
        data = json.load(file)
        DEFAULTS['release'] = data['version']
        DEFAULTS['priority'] = data['priority']

    if ARGS.cores is not None:
        print("INFO: Cores = {}".format(ARGS.cores))
        DEFAULTS['make_cores'] = ARGS.cores

    if ARGS.log is not None:
        print("INFO: Log = {}".format(ARGS.log))
        DEFAULTS['make_loglevel'] = ARGS.log

    if "/" in ARGS.branch:
        DEFAULTS['branch'] = ARGS.branch.split("/")[1]
        print("Warning: found \"/\" in branch name, changing to: {}".format(DEFAULTS['branch']))
    elif ARGS.branch is not None:
        DEFAULTS['branch'] = ARGS.branch
        print("Info: branch switched to: {}".format(DEFAULTS['branch']))

    if ARGS.command == "clean":
        clean()
    elif ARGS.command == "dirclean":
        dirclean()
    elif ARGS.command == "update":
        update()
    elif ARGS.command == "build":
        build()
    elif ARGS.command == "sign":
        sign()
    elif ARGS.command == "publish":
        publish()
    else:
        raise ValueError("Command can only be one of: clean |"+
                         " dirclean | update | build | sign | publish")


if __name__ == '__main__':
    main()
